### PR TITLE
Adding support for mouse-wheel scrolling and zooming while mandatory irregular or repeated snap points are applied

### PR DESCRIPTION
**Pull request by RBrid**
Attempting to scroll or zoom with the mouse wheel while mandatory snap points were defined was almost impossible because the current snapped location would win.

The approach to avoid this is to introduce the concept of ignored snap value. When landing on a snap value and it's not alone, that snap value becomes ignored when inertia is triggered by a mouse wheel impulse. 
The InteractionTracker.IsInertiaFromImpulse property, introduced in RS5 to address this bug, is used to determine if inertia was triggered by a mouse wheel impulse. For lower level versions where TryUpdatePositionWithAdditionalVelocity and TryUpdateScaleWithAdditionalVelocity are used, we use the new Scroller::m_isInertiaFromImpulse field.

A snap point now has an impulse applicable zone, along with the old applicable zone. The applicable zone is used by default. When inertia is from an impulse, the impulse applicable zone kicks in. The ignored snap point grows the applicable zone of its neighbors.

The snap point composition expressions now need to handle the regular case, and the inertia-from-impulse case. In those expressions, on RS5+, this.target.IsInertiaFromImpulse is leveraged. Before RS5, a boolean parameter representing Scroller::m_isInertiaFromImpulse is used. 
The expressions either use the applicable zone or the impulse applicable zone depending on the inertia type.

As few updates as possible are made when the ignored value changes, or when the inertia type changes.

### CODE REVIEW

File: dev/Scroller/SnapPointWrapper.cpp [hunk 1]
```diff
@@ -36,72 +36,168 @@ int SnapPointWrapper<T>::CombinationCount() const
...
+        previousIgnoredValue,
+        nextIgnoredValue);
 }
 
 template<typename T>
 void SnapPointWrapper<T>::Combine(SnapPointWrapper<T>* snapPointWrapper)
 {
-    winrt::SnapPointBase winrtSnapPoint = safe_cast<winrt::SnapPointBase>(m_snapPoint);
-    SnapPointBase* snapPoint = winrt::get_self<SnapPointBase>(winrtSnapPoint);
+    SnapPointBase* snapPoint = GetSnapPointFromWrapper(this);
```
Comment (StephenLPeters) [inline dev/Scroller/SnapPointWrapper.cpp]: I'd pobably just combine these two lines, assuming GetSnapPointFromWrapper doesn't ever return null, which we aren't checking for anyway

**Comment by StephenLPeters** [comment]
In general I think a lot of this code would benefit from using lambdas to instantiate local variables, there are a lot of raw pointers floating around and scope guarding them would make the code more intentional.

As far as I can tell the logic is correct, I didn't scrutinize the expressions though.

I expected to see a new test for this change. Is the updated coverage enough?

File: dev/Scroller/Scroller.cpp [hunk 2]
```diff
@@ -1832,12 +1854,38 @@ void Scroller::SetupSnapPoints(
...
-    FixSnapPointRanges(snapPointsSet);
-
     if (!m_interactionTracker)
     {
         EnsureInteractionTracker();
     }
+
+    if (m_state == winrt::InteractionState::Idle)
+    {
+        double ignoredValue = 0.0;
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/Scroller.cpp]: Done.

File: dev/Scroller/Scroller.cpp [hunk 3]
```diff
@@ -1941,15 +1990,173 @@ void Scroller::FixSnapPointRanges(
...
+            forImpulseOnly);
+    }
+}
+
+template <typename T>
+void Scroller::UpdateSnapPointsIgnoredValue(
+    std::set<std::shared_ptr<SnapPointWrapper<T>>, SnapPointWrapperComparator<T>>* snapPointsSet,
+    ScrollerDimension dimension)
+{
+    double newIgnoredValue = 0.0;
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/Scroller.cpp]: Done.

File: dev/Scroller/Scroller.cpp [hunk 4]
```diff
@@ -1941,15 +1990,173 @@ void Scroller::FixSnapPointRanges(
...
+        MUX_ASSERT(false);
+    }
+
+    if (UpdateSnapPointsIgnoredValue(snapPointsSet, newIgnoredValue))
+    {
+        // The ignored snap point value has changed.
+        UpdateSnapPointsRanges(snapPointsSet, true /*forImpulseOnly*/);
+
+        winrt::Compositor compositor = m_interactionTracker.Compositor();
+        winrt::IVector<winrt::InteractionTrackerInertiaModifier> modifiers = winrt::make<Vector<winrt::InteractionTrackerInertiaModifier>>();
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/Scroller.cpp]: I skipped that - I think a lambda would be overkill here.

File: dev/Scroller/Scroller.cpp [hunk 5]
```diff
@@ -1941,15 +1990,173 @@ void Scroller::FixSnapPointRanges(
...
+
+        for (std::shared_ptr<SnapPointWrapper<T>> snapPointWrapper : *snapPointsSet)
+        {
+            winrt::InteractionTrackerInertiaRestingValue modifier = winrt::InteractionTrackerInertiaRestingValue::Create(compositor);
+            winrt::ExpressionAnimation conditionExpressionAnimation = nullptr;
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/Scroller.cpp]: Cool. Thanks.

File: dev/Scroller/Scroller.cpp [hunk 6]
```diff
@@ -1941,15 +1990,173 @@ void Scroller::FixSnapPointRanges(
...
+
+    if (UpdateSnapPointsIgnoredValue(snapPointsSet, newIgnoredValue))
+    {
+        // The ignored snap point value has changed.
+        UpdateSnapPointsRanges(snapPointsSet, true /*forImpulseOnly*/);
+
+        winrt::Compositor compositor = m_interactionTracker.Compositor();
+        winrt::IVector<winrt::InteractionTrackerInertiaModifier> modifiers = winrt::make<Vector<winrt::InteractionTrackerInertiaModifier>>();
+
+        for (std::shared_ptr<SnapPointWrapper<T>> snapPointWrapper : *snapPointsSet)
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/Scroller.cpp]: Done.

File: dev/Scroller/Scroller.cpp [hunk 7]
```diff
@@ -6826,22 +7047,37 @@ winrt::InteractionTrackerInertiaRestingValue Scroller::GetInertiaRestingValue(
...
 
     return modifier;
 }
 
-bool Scroller::IsLoaded()
+// Relies on InteractionTracker.IsInertiaFromImpulse starting with RS5,
+// returns the replacement field m_isInertiaFromImpulse otherwise.
+bool Scroller::IsInertiaFromImpulse() const
+{
+    if (SharedHelpers::IsRS5OrHigher() && m_interactionTracker)
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/Scroller.cpp]: Done.

File: dev/Scroller/ScrollerSnapPoint.cpp [hunk 8]
```diff
@@ -154,42 +207,93 @@ double ScrollSnapPoint::Value()
...
+        scale.data());
+    winrt::hstring scaledMinImpulseApplicableRange = StringUtil::FormatString(
+        scaledValue.data(),
+        s_minImpulseApplicableValue.data(),
+        scale.data());
+    winrt::hstring scaledMaxImpulseApplicableRange = StringUtil::FormatString(
+        scaledValue.data(),
+        s_maxImpulseApplicableValue.data(),
         scale.data());
     winrt::hstring expression = StringUtil::FormatString(
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/ScrollerSnapPoint.cpp]: I skipped that - I think a lambda would be overkill here.

File: dev/Scroller/ScrollerSnapPoint.cpp [hunk 9]
```diff
@@ -207,6 +311,24 @@ std::tuple<double, double> ScrollSnapPoint::DetermineActualApplicableZone(
...
 }
 
+std::tuple<double, double> ScrollSnapPoint::DetermineActualImpulseApplicableZone(
+    SnapPointBase* previousSnapPoint,
+    SnapPointBase* nextSnapPoint,
+    double currentIgnoredValue,
+    double previousIgnoredValue,
+    double nextIgnoredValue)
+{
+    return std::tuple<double, double>{
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/ScrollerSnapPoint.cpp]: Done.

File: dev/Scroller/ScrollerSnapPoint.cpp [hunk 10]
```diff
@@ -432,102 +629,169 @@ double RepeatedScrollSnapPoint::End()
...
-    restingPointExpressionAnimation.SetScalarParameter(L"first", static_cast<float>(DetermineFirstRepeatedSnapPointValue()));
+        L"((Abs(it.%2!s!/it.Scale-((Floor((it.%2!s!/it.Scale-fst)/itv)*itv)+fst))>=Abs(it.%2!s!/it.Scale-((Ceil((it.%2!s!/it.Scale-fst)/itv)*itv)+fst)))&&(((Ceil((it.%2!s!/it.Scale-fst)/itv)*itv)+fst)<=(%1!s!?iEnd:end)))?(%1!s!?(((Ceil((it.%2!s!/it.Scale-fst)/itv)*itv)+fst)==iIgn?((iIgn==fst?fst*it.Scale:(iIgn-itv)*it.Scale)):((Ceil((it.%2!s!/it.Scale-fst)/itv)*itv)+fst)*it.Scale):((Ceil((it.%2!s!/it.Scale-fst)/itv)*itv)+fst)*it.Scale):(%1!s!?(((Floor((it.%2!s!/it.Scale-fst)/itv)*itv)+fst)==iIgn?(iIgn+itv<=(%1!s!?iEnd:end)?(iIgn+itv)*it.Scale:iIgn*it.Scale):((Floor((it.%2!s!/it.Scale-fst)/itv)*itv)+fst)*it.Scale):((Floor((it.%2!s!/it.Scale-fst)/itv)*itv)+fst)*it.Scale)",
+        isInertiaFromImpulseExpression.data(),
+        target.data());
+
+    SCROLLER_TRACE_VERBOSE(nullptr, TRACE_MSG_METH_STR, METH_NAME, this, expression.c_str());
+
+    winrt::ExpressionAnimation restingPointExpressionAnimation = interactionTracker.Compositor().CreateExpressionAnimation(expression);
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/ScrollerSnapPoint.cpp]: Done.

File: dev/Scroller/ScrollerSnapPoint.cpp [hunk 11]
```diff
@@ -432,102 +629,169 @@ double RepeatedScrollSnapPoint::End()
...
-    conditionExpressionAnimation.SetScalarParameter(L"end", static_cast<float>(ActualEnd()));
-    conditionExpressionAnimation.SetScalarParameter(L"applicableRange", static_cast<float>(m_specifiedApplicableRange));
-    conditionExpressionAnimation.SetScalarParameter(L"first", static_cast<float>(DetermineFirstRepeatedSnapPointValue()));
+        L"((!%1!s! && it.%2!s! / it.Scale >= stt && it.%2!s! / it.Scale <= end) || (%1!s! && it.%2!s! / it.Scale >= iStt && it.%2!s! / it.Scale <= iEnd)) && (((Floor((it.%2!s! / it.Scale - fst) / itv) * itv) + fst + aRg >= it.%2!s! / it.Scale) || (((Ceil((it.%2!s! / it.Scale - fst) / itv) * itv) + fst - aRg <= it.%2!s! / it.Scale) && ((Ceil((it.%2!s! / it.Scale - fst) / itv) * itv) + fst <= (%1!s! ? iEnd : end))))",
+        isInertiaFromImpulseExpression.data(),
+        target.data());
+
+    SCROLLER_TRACE_VERBOSE(nullptr, TRACE_MSG_METH_STR, METH_NAME, this, expression.c_str());
+
+    winrt::ExpressionAnimation conditionExpressionAnimation = interactionTracker.Compositor().CreateExpressionAnimation(expression);
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/ScrollerSnapPoint.cpp]: I think a lambda would be overkill here.

File: dev/Scroller/ScrollerSnapPoint.cpp [hunk 12]
```diff
@@ -553,6 +817,24 @@ std::tuple<double, double> RepeatedScrollSnapPoint::DetermineActualApplicableZon
...
 }
 
+std::tuple<double, double> RepeatedScrollSnapPoint::DetermineActualImpulseApplicableZone(
+    SnapPointBase* previousSnapPoint,
+    SnapPointBase* nextSnapPoint,
+    double currentIgnoredValue,
+    double previousIgnoredValue,
+    double nextIgnoredValue)
+{
+    return std::tuple<double, double>{
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/ScrollerSnapPoint.cpp]: Done.

File: dev/Scroller/ScrollerSnapPoint.cpp [hunk 13]
```diff
@@ -755,34 +1118,79 @@ double ZoomSnapPoint::Value()
...
+    winrt::InteractionTracker const& interactionTracker,
     winrt::hstring const& target,
-    winrt::hstring const& scale)
+    winrt::hstring const& scale,
+    bool isInertiaFromImpulse)
 {
-    winrt::ExpressionAnimation restingPointExpressionAnimation = compositor.CreateExpressionAnimation(L"snapPointValue");
+    SCROLLER_TRACE_VERBOSE(nullptr, TRACE_MSG_METH, METH_NAME, this);
+
+    winrt::ExpressionAnimation restingPointExpressionAnimation = interactionTracker.Compositor().CreateExpressionAnimation(s_snapPointValue);
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/ScrollerSnapPoint.cpp]: Done.

File: dev/Scroller/ScrollerSnapPoint.cpp [hunk 14]
```diff
@@ -755,34 +1118,79 @@ double ZoomSnapPoint::Value()
...
+{
+    SetScalarParameter(conditionExpressionAnimation, s_minImpulseApplicableValue, static_cast<float>(std::get<0>(actualImpulseApplicableZone)));
+    SetScalarParameter(conditionExpressionAnimation, s_maxImpulseApplicableValue, static_cast<float>(std::get<1>(actualImpulseApplicableZone)));
+}
+
+void ZoomSnapPoint::UpdateRestingPointExpressionAnimationForImpulse(
+    winrt::ExpressionAnimation const& restingValueExpressionAnimation,
+    double ignoredValue,
+    std::tuple<double, double> actualImpulseApplicableZone) const
+{
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/ScrollerSnapPoint.cpp]: Done.

File: dev/Scroller/ScrollerSnapPoint.cpp [hunk 15]
```diff
@@ -798,6 +1206,24 @@ std::tuple<double, double> ZoomSnapPoint::DetermineActualApplicableZone(
...
 }
 
+std::tuple<double, double> ZoomSnapPoint::DetermineActualImpulseApplicableZone(
+    SnapPointBase* previousSnapPoint,
+    SnapPointBase* nextSnapPoint,
+    double currentIgnoredValue,
+    double previousIgnoredValue,
+    double nextIgnoredValue)
+{
+    return std::tuple<double, double>{
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/ScrollerSnapPoint.cpp]: Done.

File: dev/Scroller/ScrollerSnapPoint.cpp [hunk 16]
```diff
@@ -1104,6 +1703,24 @@ std::tuple<double, double> RepeatedZoomSnapPoint::DetermineActualApplicableZone(
...
 }
 
+std::tuple<double, double> RepeatedZoomSnapPoint::DetermineActualImpulseApplicableZone(
+    SnapPointBase* previousSnapPoint,
+    SnapPointBase* nextSnapPoint,
+    double currentIgnoredValue,
+    double previousIgnoredValue,
+    double nextIgnoredValue)
+{
+    return std::tuple<double, double>{
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/ScrollerSnapPoint.cpp]: Done.

File: dev/Scroller/SnapPointWrapper.cpp [hunk 17]
```diff
@@ -36,72 +36,168 @@ int SnapPointWrapper<T>::CombinationCount() const
...
-        scale);
+        scale,
+        isInertiaFromImpulse);
+
+    return m_conditionExpressionAnimation;
+}
+
+// Invoked when the InteractionTracker reaches the Idle State and a new ignored value may have to be set.
+template<typename T>
+void SnapPointWrapper<T>::GetUpdatedExpressionAnimationsForImpulse(
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/SnapPointWrapper.cpp]: Done.

File: dev/Scroller/SnapPointWrapper.cpp [hunk 18]
```diff
@@ -36,72 +36,168 @@ int SnapPointWrapper<T>::CombinationCount() const
...
+        m_actualImpulseApplicableZone);
+
+    *conditionalExpressionAnimation = m_conditionExpressionAnimation;
+    *restingPointExpressionAnimation = m_restingValueExpressionAnimation;
+}
+
+// Invoked on pre-RS5 versions when Scroller::m_isInertiaFromImpulse changed
+// and the 'iIFI' boolean parameters need to be updated.
+template<typename T>
+void SnapPointWrapper<T>::GetUpdatedExpressionAnimationsForImpulse(
```
[the remark this answers is not in the document]
Comment (RBrid) [reply dev/Scroller/SnapPointWrapper.cpp]: Done.

File: dev/Scroller/SnapPointWrapper.cpp (hunk 1 above)
Comment (RBrid) [reply dev/Scroller/SnapPointWrapper.cpp]: Done.

**Comment by RBrid** [comment]
Thanks for the review.
> I expected to see a new test for this change. Is the updated coverage enough?

I think the changes made in ScrollerTestsWithInputHelper.cs are fine.

File: dev/Scroller/Scroller.cpp [hunk 19]
```diff
@@ -6827,22 +7035,39 @@ winrt::InteractionTrackerInertiaRestingValue Scroller::GetInertiaRestingValue(
...
-bool Scroller::IsLoaded()
+// Relies on InteractionTracker.IsInertiaFromImpulse starting with RS5,
+// returns the replacement field m_isInertiaFromImpulse otherwise.
+bool Scroller::IsInertiaFromImpulse() const
+{
+    if (m_interactionTracker)
+    {
+        winrt::IInteractionTracker4 interactionTracker4 = m_interactionTracker;
+
+        return interactionTracker4 ? interactionTracker4.IsInertiaFromImpulse() : m_isInertiaFromImpulse;
```
Comment (StephenLPeters) [inline dev/Scroller/Scroller.cpp]: The result of the instantiated interactionTracker4 will be null if m_interactionTracker is null, or if it is not on a version where the interface is present so we can move the declaration to the if check.

if(winrt::IInteractionTracker4 interactionTracker4 = m_interactionTracker)
{
    return interactionTracker4.IsInertiaFromImpulse();
}

**Pull request merged**